flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -2159,138 +2159,267 @@ struct CoordinateOpConversion
+        auto idxTy = lowerTy().indexType();
+        mlir::Value off = genConstantIndex(loc, idxTy, rewriter, 0);
+        for (unsigned index = i, lastIndex = i + arrTy.getDimension();
+             index < lastIndex; ++index) {
+          mlir::Value stride =
+              loadStrideFromBox(loc, operands[0], index - i, rewriter);
+          auto sc = rewriter.create<mlir::LLVM::MulOp>(loc, idxTy,
+                                                       operands[index], stride);
+          off = rewriter.create<mlir::LLVM::AddOp>(loc, idxTy, sc, off);
         }
+        auto voidPtrBase =
+            rewriter.create<mlir::LLVM::BitcastOp>(loc, voidPtrTy, resultAddr);
+        SmallVector<mlir::Value> args{voidPtrBase, off};
+        resultAddr = rewriter.create<mlir::LLVM::GEPOp>(loc, voidPtrTy, args);
+        i += arrTy.getDimension() - 1;
+        currentObjTy = arrTy.getEleTy();
+      } else if (auto recTy = currentObjTy.dyn_cast<fir::RecordType>()) {
+        auto recRefTy =
+            mlir::LLVM::LLVMPointerType::get(lowerTy().convertType(recTy));
+        mlir::Value nxtOpnd = operands[i];
+        auto memObj =
+            rewriter.create<mlir::LLVM::BitcastOp>(loc, recRefTy, resultAddr);
+        llvm::SmallVector<mlir::Value> args = {memObj, c0, nxtOpnd};
+        currentObjTy = recTy.getType(getFieldNumber(recTy, nxtOpnd));
+        auto llvmCurrentObjTy = lowerTy().convertType(currentObjTy);
+        auto gep = rewriter.create<mlir::LLVM::GEPOp>(
+            loc, mlir::LLVM::LLVMPointerType::get(llvmCurrentObjTy), args);
+        resultAddr =
+            rewriter.create<mlir::LLVM::BitcastOp>(loc, voidPtrTy, gep);
+      } else {
+        fir::emitFatalError(loc, "unexpected type in coordinate_of");
       }
-      rewriter.replaceOpWithNewOp<mlir::LLVM::BitcastOp>(coor, ty, addr);
-      return success();
     }
-    if (!hasSubdimension)
-      columnIsDeferred = true;
 
-    if (!validCoordinate(cpnTy, operands.drop_front(1)))
-      return mlir::emitError(loc, "coordinate has incorrect dimension");
+    rewriter.replaceOpWithNewOp<mlir::LLVM::BitcastOp>(coor, ty, resultAddr);
+    return success();
+  }
+
+  mlir::LogicalResult
+  doRewriteRefOrPtr(fir::CoordinateOp coor, mlir::Type ty,
+                    mlir::ValueRange operands, mlir::Location loc,
+                    mlir::ConversionPatternRewriter &rewriter) const {
+    mlir::Type baseObjectTy = coor.getBaseType();
+
+    mlir::Type currentObjTy = fir::dyn_cast_ptrOrBoxEleTy(baseObjectTy);
+    bool hasSubdimension = hasSubDimensions(currentObjTy);
+    bool columnIsDeferred = !hasSubdimension;
+
+    if (!supportedCoordinate(currentObjTy, operands.drop_front(1))) {
+      TODO(loc, "unsupported combination of coordinate operands");
+    }
 
-    // if arrays has known shape
     const bool hasKnownShape =
-        arraysHaveKnownShape(cpnTy, operands.drop_front(1));
+        arraysHaveKnownShape(currentObjTy, operands.drop_front(1));
```

Comment:
`cpnTy` means the component type.

Reply:
> `cpnTy` means the component type.

Thanks for clarifying, that wasn't clear when I was reading this code. I would really prefer something more descriptive than `cpnTy`, perhaps `currentObjTy` or `componentTy`?

To me, `currentObjTy` (or `currentComponentTy`) would be a bit more accurate than e.g. `componentTy`:
* `currentObjTy` (or `currentComponentTy`) indicates that this type is likely to be updated (indeed, it is updated on every iteration)
* `componentTy` suggests that this type is static (i.e. that it's always referring to the same component).

This is a light preference. What's yours?

Reply:
Well, it isn't the type of the object. As you say as the object is traversed "inwards", the component type is updated to reflect (insert drum roll), the type of the component. :)  Since components can be derived recursively, the component type is relative to the current logic position in the path. Using the word "current" everywhere seems redundant, after all what else would the component type be other than the type of the current logic position? We don't really care what the component 4 levels previous was and we don't know what it might be 5 iterations further down. So, "current" seems like 7 characters of line noise to me.

Reply:
Thanks for the explanation! Our mental models are slightly different, but I like yours more (I also dislike "current" in the variable name). This brings us to `componentTy`, but `cpnTy` would be more compact! Let me switch back to `cpnTy` and simply add a comment so that `cpnTy` is not so enigmatic to people unfamiliar with this code.